### PR TITLE
Merge SIMD proposal

### PR DESCRIPTION
**Pull request by ngzhian**
SIMD is [phase 5](https://github.com/WebAssembly/simd/issues/507), merge all the changes back into main spec.

### CODE REVIEW

**Comment by Ms2ger** [comment]
Could you squash the history into one commit, like #1144 for example?

File: document/js-api/index.bs
Comment (Ms2ger) [inline document/js-api/index.bs:901]:
```suggestion
```

File: test/js-api/global/constructor.any.js
Comment (Ms2ger) [inline test/js-api/global/constructor.any.js:170]:
```suggestion
  assert_throws_js(TypeError, () => new WebAssembly.Global(argument));
```

**Comment by ngzhian** [comment]
> Could you squash the history into one commit, like #1144 for example?

I can use the "Squash and merge" button when merging this PR, does that work?

**Comment by Ms2ger** [comment]
I suppose that works, yes. Thanks.

**Review by dtig — APPROVED** [review]
Thanks for putting this together! Please wait for a review from @rossberg as well.

File: meetings/meeting-4-24-19.md
Comment (rossberg) [inline meetings/meeting-4-24-19.md:1]: This file does not seem to belong here?

**Review by rossberg — COMMENTED** [review]
Thanks! This PR is so large that the GitHub UI broke down on me...

I focussed on the places where this PR changes existing spec text and discovered a few small oversights that should be easy to fix.

Perhaps more seriously, I noticed only now that the implementation of packed int types in the interpreter is probably insufficient/incorrect as is. Better to fix this before landing.

**Comment by ngzhian** [comment]
> Thanks! This PR is so large that the GitHub UI broke down on me...
> 
> I focussed on the places where this PR changes existing spec text and discovered a few small oversights that should be easy to fix.
> 
> Perhaps more seriously, I noticed only now that the implementation of packed int types in the interpreter is probably insufficient/incorrect as is. Better to fix this before landing.

Do you prefer I make changes as commits to this PR, or make changes to upstream simd repo, then update this PR? (the end result will be the same, a single commit to merge simd, but thinking which would be easier for your to review)

File: interpreter/exec/ixx.ml
Comment (ngzhian) [reply interpreter/exec/ixx.ml:206]: We don't have that problem because all the SIMD operations pass through `to_lanes` and `of_lanes`, which strips the junk bits and appropriately extends to i32, so later on the shr_s works correctly.

I.e. i8.ml itself is broken, but it's usage together with v128.ml is fine.

`shr_u` requires this fix (i think) because it is used in `of_string`.

File: document/core/intro/overview.rst
Comment (ngzhian) [reply document/core/intro/overview.rst:32]: This seems like it should be added to the spec first, instead of as part of this PR?

Comment (rossberg) [reply document/core/intro/overview.rst:32]: Fair enough, I created #1394.

File: interpreter/exec/ixx.ml
Comment (rossberg) [reply interpreter/exec/ixx.ml:206]: Ah, I see, good. But I still feel uncomfortable with having broken implementations of the I8/I16 modules themselves. It is likely that somebody else is going to try using them in the future for purposes where this problem is not masked.

Comment (ngzhian) [reply interpreter/exec/ixx.ml:206]: Fixed in https://github.com/WebAssembly/simd/pull/528

File: interpreter/exec/ixx.ml
Comment (ngzhian) [reply interpreter/exec/ixx.ml:37]: https://github.com/WebAssembly/simd/pull/530

**Comment by ngzhian** [comment]
This is good to review, thanks!

**Review by rossberg — APPROVED** [review]
Excellent, thanks a lot!

"+100,917 -690" -- just wow.

**Comment by ngzhian** [comment]
For fun, lines of changes, breakdown by filetype:

wast: 81666 (most of which are generated by py scripts)
py: 9866 (tests)
ml: 3204 (ref interpreter)
rst: 3763 (spec text)

File: document/core/exec/runtime.rst
Comment (codefromthecrypt) [inline document/core/exec/runtime.rst:39]: is there any documentation we can refer to that clarifies why this is v128.const and not i128.const for declaring an i128? Or is this possibly a bug in the spec?

Comment (codefromthecrypt) [reply document/core/exec/runtime.rst:39]: I guess it is a shortcut rather than declaring i128.const and then another vector instruction that produces it? nevertheless is there a plan or explicit plan against i128.const?

Comment (rossberg) [reply document/core/exec/runtime.rst:39]: PS: As for documentation, that is essentially said in https://webassembly.github.io/spec/core/syntax/values.html#vectors

**Pull request merged**
